main.go

```diff
@@ -111,8 +116,15 @@ func run(c *cli.Context) error {
 		go func() { startchan <- struct{}{} }()
 		file := c.String("file")
 		configchan := make(chan bool)
+		// XXX Unwatchable Puppet sources should get events from a timer,
+		// with the same interval as the Puppet agent
+		// (hint: puppet config print runinterval)
 		if !c.Bool("no-watch") {
-			configchan = ConfigWatch(file)
+			if c.IsSet("file") {
+				configchan = ConfigWatch(file)
+			} else if strings.HasSuffix(c.String("puppet"), ".pp") {
+				configchan = ConfigWatch(c.String("puppet"))
```

Comment:
While this is clever, and you get style points, I'd actually like to argue against this, because it's not consistent with well-known puppet behaviour. As a bonus for removing it, we also don't expose the parsing logic in more than one place too.

Instead, as you XXX mentions (and I whole heartedly agree) we should just implement this :) One catch would be to check that the converge timeout isn't longer than the puppet timeout or it will never end. Maybe add a warning in this case.